src/components/InfluenceChart/index.js

```diff
@@ -0,0 +1,751 @@
+  person: PersonAbstract | PersonDetail,
+  getId: () => string,
+}
+
+
+type TLink = {
+  source: PersonNode,
+  middle: InvisibleNode,
+  target: PersonNode,
+}
+
+
+class TGraph {
+  nodes: { [SubjectId]: InvisibleNode | PersonNode }
+  links: Array<TLink>
+  focus: PersonNode
+
+  constructor(focus: PersonDetail) {
+    this.nodes = {}
+    this.links = []
+
+    this.setFocus(focus)
+  }
+
+  setFocus(person: PersonDetail): PersonNode {
+    const node = this.addPerson(person)
+    this.focus = node && node.type === 'PersonNode' ? node : this.focus
+    return node
+  }
+
+  addNode(pn: PersonNode): void {
+    this.nodes[pn.getId()] = pn
+  }
+
+  addPerson(person: PersonAbstract | PersonDetail): PersonNode {
+    const p = this.nodes[person.id]
+    if (p != null && p.type === 'PersonNode' && p.person.type === person.type) {
+      return p
+    }
+
+    const node = {
+      type: 'PersonNode',
+      x: 0,
+      y: 0,
+      vx: 0,
+      vy: 0,
+      person,
+      getId: () => person.id,
+    }
+    this.addNode(node)
+    return node
+  }
+
+  removePerson(person: SubjectId | PersonAbstract | PersonDetail): void {
+    /* remove the person, all links going to or leaving that person, and the
+     * middle nodes for thos e links */
+    const personId = typeof person === 'string' ? person : person.id
+
+    delete this.nodes[personId]
+
+    const removeLinks = fp.filter(
+      (l: TLink): bool => l.source.getId() === personId || l.target.getId() === personId,
+      this.links,
+    )
+
+    this.links = fp.filter(
+      (l: TLink): bool => l.source.getId() !== personId && l.target.getId() !== personId,
+      this.links,
+    )
+
+    removeLinks.forEach((l: TLink): void => {
+      delete this.nodes[l.middle.getId()]
+    })
+  }
+
+  createLink(source: PersonAbstract | PersonDetail, target: PersonAbstract | PersonDetail): ?TLink {
+    if (source === target) {
+      return null
+    }
+    const sourceNode = this.addPerson(source)
+    const targetNode = this.addPerson(target)
+    const middle = {
+      type: 'InvisibleNode',
+      x: 0,
+      y: 0,
+      vx: 0,
+      vy: 0,
+      getId: () => `${source.id} - ${target.id}`,
+    }
+    const link = { source: sourceNode, middle, target: targetNode }
+
+    this.nodes[middle.getId()] = middle
+    this.links.push(link)
+    return link
+  }
+
+  getNodes(): Array<InvisibleNode | PersonNode> {
+    return fp.values(this.nodes)
+  }
+
+  getVisibleNodes(): Array<PersonNode> {
+    return fp.filter(n => n.type === 'PersonNode')(fp.values(this.nodes))
+  }
+
+  getLinks(): Array<TLink> {
+    return this.links
+  }
+
+  getLinkSegments(): Array<LinkSegment> {
+    return fp.flatten(fp.map(link => [
+      { source: link.source, target: link.middle },
+      { source: link.middle, target: link.target },
+    ])(this.links))
+  }
+}
+
+
+type Dimensions = { width: number, height: number }
+
+const convertToSafeDomId = (str: string): string => str.replace(/(%20| |\.)/g, '_')
```

Comment:
Can this somehow be turned into a data type and encapsulated that way? It feels really weird to have it here. Maybe just utils.js.